src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -545,14 +625,41 @@ private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
             return new ValueTask<int>(vts, vts.Version);
         }
 
-        private static NativeOverlapped GetNativeOverlapped(long fileOffset)
+        private static unsafe NativeOverlapped* GetNativeOverlappedForSynchronousOperation(SafeFileHandle handle, long fileOffset, NativeOverlapped stackAllocated = default)
         {
-            NativeOverlapped nativeOverlapped = default;
+            NativeOverlapped* result;
+            if (handle.IsAsync)
+            {
+                handle.EnsureThreadPoolBindingInitialized();
+
+                // After SafeFileHandle is bound to ThreadPool, we need to use ThreadPoolBinding
+                // to allocate a native overlapped and provide a valid callback.
+                // Since we really don't care about the callback (because this is sync IO for async handle)
+                // and we are going to wait on WaitHandle anyway, we pass ThreadPoolBinding as a state.
+                // The callback is going to use it to free the native overlapped.
+                result = handle.ThreadPoolBinding!.AllocateNativeOverlapped(s_callback, handle.ThreadPoolBinding, null);
+            }
+            else
+            {
+                result = &stackAllocated;
```

Comment:
I would expect this to result in sporadic crashes or AVs.  This is taking the address of a local (the parameter), and returning that address; that memory can get stomped on / overwritten (or even freed if the original thread were to go away).